Use `#with_raw_connection` in `#quote_string` to retry connection errors

adrianna-chang-shopify opened the pull request:
### Motivation / Background

Follow-up to the work done in #44576, #44591, #46046

MySQL and PostgreSQL both rely on a connection to perform string quoting. If the raw connection ends up being closed (network errors, server-side disconnect, etc.), instead of immediately raising a connection error, we can instead reconnect and retry the quoting by going through `#with_raw_connection`.

Shopify has historically patched `#quote_string` to rescue and retry connection-related exceptions. Often `#quote` is called before querying the db, so if we've lost connection to the db, we'll fail on `#quote_string` and raise without making it to the query to be retried. We were using the same logic in our patch for both `#quote_string` and `#execute`, so now that `#execute` will reconnect and retry upstream, we're motivated to have `#quote_string` reuse that logic so that our patch can go away completely.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] There are no typos in commit messages and comments.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Feature branch is up-to-date with `main` (if not - rebase it).
* [X] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [X] Tests are added if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [X] PR is not in a draft state.
* [X] CI is passing.

byroot commented on activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb:
```diff
@@ -41,9 +41,30 @@ def quote(value) # :nodoc:
+        rescue PG::ConnectionBad => error
+          translated_exception = translate_exception(error, message: error.message, sql: "<escape>", binds: [])
+
+          now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+          query_time = now - start_time
+          retry_deadline_exceeded = retry_deadline && query_time > retry_deadline
+
+          if !retry_deadline_exceeded && retries_available > 0
+            retries_available -= 1
+
+            if retryable_connection_error?(translated_exception)
+              reconnect!
+              retry
+            end
+          end
```
Maybe we should extract this code in a private method in `AbstractAdapter`? I feel like we're coyping it a lot, and it's non-trivial with the deadline logic.

adrianna-chang-shopify replied:
Yeah, I was thinking the same thing -- although the code does vary quite a bit between this and the retry logic in `#with_raw_connection`, although there's probably overlap that would make sense to extract.

To keep things simple, I'll wait until https://github.com/rails/rails/pull/46046 is merged so I can clean up the deadline logic to match the changes I just made, and then I'll look into extracting this 😄

matthewd replied:
> the code does vary quite a bit between this and the retry logic in `#with_raw_connection`

Should it?

I came here to claim that [if we want retries] this method should be `with_raw_connection { |c| c.escape(s) }`

adrianna-chang-shopify replied:
That makes this much simpler to solve 🤦‍♀️ Thanks Matthew!

adrianna-chang-shopify committed:
Use connection from `#with_raw_connection` in `#quote_string`.

This ensures we hook into the retry logic for connection-related exceptions
that `#with_raw_connection` provides.

adrianna-chang-shopify commented on activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:
```diff
@@ -97,10 +97,11 @@ def error_number(exception)
-        any_raw_connection.escape(string)
-      rescue Mysql2::Error => error
-        raise translate_exception(error, message: error.message, sql: "<escape>", binds: [])
+        with_raw_connection(allow_retry: true, uses_transaction: false) do |connection|
```
I'm assuming `uses_transaction: false` here is what we want, since it seems like we shouldn't need transactions to be materialized in order to do string quoting, but please correct me if I'm wrong.

byroot replied:
That's correct
